### ISSUE TITLE
date[("get" + method)] is not a function

### ISSUE THREAD

**Issue by michaelimstepf** [intljusticemission/react-big-calendar#163]
I'm a Rails dev and new to Node/Webpack/React. When I switch between month/week/day view, the app often crashes and I get this error in the console: `index.js?07ad:209 Uncaught TypeError: date[("get" + method)] is not a function`.

```jsx
<BigCalendar
  defaultDate={new Date()}
  defaultView='week'
  events={this.state.events}
/>
```

I wonder whether this has to do with how webpack is configured and compiles the files? Here's the file that causes the error, the problem is on line 209:

```javascript
var MILI    = 'milliseconds'
  , SECONDS = 'seconds'
  , MINUTES = 'minutes'
  , HOURS   = 'hours'
  , DAY     = 'day'
  , WEEK    = 'week'
  , MONTH   = 'month'
  , YEAR    = 'year'
  , DECADE  = 'decade'
  , CENTURY = 'century';

var dates = module.exports = {

  add: function(date, num, unit) {
    date = new Date(date)

    switch (unit){
      case MILI:
      case SECONDS:
      case MINUTES:
      case HOURS:
      case YEAR:
        return dates[unit](date, dates[unit](date) + num)
      case DAY:
        return dates.date(date, dates.date(date) + num)
      case WEEK:
        return dates.date(date, dates.date(date) + (7 * num)) 
      case MONTH:
        return monthMath(date, num)
      case DECADE:
        return dates.year(date, dates.year(date) + (num * 10))
      case CENTURY:
        return dates.year(date, dates.year(date) + (num * 100))
    }

    throw new TypeError('Invalid units: "' + unit + '"')
  },

  subtract: function(date, num, unit) {
    return dates.add(date, -num, unit)
  },

  startOf: function(date, unit, firstOfWeek) {
    date = new Date(date)

    switch (unit) {
      case 'century':
      case 'decade':
      case 'year':
          date = dates.month(date, 0);
      case 'month':
          date = dates.date(date, 1);
      case 'week':
      case 'day':
          date = dates.hours(date, 0);
      case 'hours':
          date = dates.minutes(date, 0);
      case 'minutes':
          date = dates.seconds(date, 0);
      case 'seconds':
          date = dates.milliseconds(date, 0);
    }

    if (unit === DECADE) 
      date = dates.subtract(date, dates.year(date) % 10, 'year')
    
    if (unit === CENTURY) 
      date = dates.subtract(date, dates.year(date) % 100, 'year')

    if (unit === WEEK) 
      date = dates.weekday(date, 0, firstOfWeek);

    return date
  },


  endOf: function(date, unit, firstOfWeek){
    date = new Date(date)
    date = dates.startOf(date, unit, firstOfWeek)
    date = dates.add(date, 1, unit)
    date = dates.subtract(date, 1, MILI)
    return date
  },

  eq:  createComparer(function(a, b){ return a === b }),
  neq: createComparer(function(a, b){ return a !== b }),
  gt:  createComparer(function(a, b){ return a > b }),
  gte: createComparer(function(a, b){ return a >= b }),
  lt:  createComparer(function(a, b){ return a < b }),
  lte: createComparer(function(a, b){ return a <= b }),

  min: function(){
    return new Date(Math.min.apply(Math, arguments))
  },

  max: function(){
    return new Date(Math.max.apply(Math, arguments))
  },
  
  inRange: function(day, min, max, unit){
    unit = unit || 'day'

    return (!min || dates.gte(day, min, unit))
        && (!max || dates.lte(day, max, unit))
  },

  milliseconds:   createAccessor('Milliseconds'),
  seconds:        createAccessor('Seconds'),
  minutes:        createAccessor('Minutes'),
  hours:          createAccessor('Hours'),
  day:            createAccessor('Day'),
  date:           createAccessor('Date'),
  month:          createAccessor('Month'),
  year:           createAccessor('FullYear'),

  decade: function (date, val) {
    return val === undefined 
      ? dates.year(dates.startOf(date, DECADE))
      : dates.add(date, val + 10, YEAR);
  },

  century: function (date, val) {
    return val === undefined 
      ? dates.year(dates.startOf(date, CENTURY))
      : dates.add(date, val + 100, YEAR);
  },

  weekday: function (date, val, firstDay) {
      var weekday = (dates.day(date) + 7 - (firstDay || 0) ) % 7;

      return val === undefined 
        ? weekday 
        : dates.add(date, val - weekday, DAY);
  },

  diff: function (date1, date2, unit, asFloat) {
    var dividend, divisor, result;

    switch (unit) {
      case MILI:
      case SECONDS:
      case MINUTES:
      case HOURS:
      case DAY:
      case WEEK:
        dividend = date2.getTime() - date1.getTime(); break;
      case MONTH:
      case YEAR:
      case DECADE:
      case CENTURY:
        dividend = (dates.year(date2) - dates.year(date1)) * 12 + dates.month(date2) - dates.month(date1); break;
      default:
        throw new TypeError('Invalid units: "' + unit + '"');
    }

    switch (unit) {
      case MILI:
          divisor = 1; break;
      case SECONDS:
          divisor = 1000; break;
      case MINUTES:
          divisor = 1000 * 60; break;
      case HOURS:
          divisor = 1000 * 60 * 60; break;
      case DAY:
          divisor = 1000 * 60 * 60 * 24; break;
      case WEEK:
          divisor = 1000 * 60 * 60 * 24 * 7; break;
      case MONTH:
          divisor = 1; break;
      case YEAR:
          divisor = 12; break;
      case DECADE:
          divisor = 120; break;
      case CENTURY:
          divisor = 1200; break;
      default:
        throw new TypeError('Invalid units: "' + unit + '"');
    }

    result = dividend / divisor;

    return asFloat ? result : absoluteFloor(result);
  }
};

function absoluteFloor(number) {
  return number < 0 ? Math.ceil(number) : Math.floor(number);
}

function monthMath(date, val){
  var current = dates.month(date)
    , newMonth  = (current + val);

    date = dates.month(date, newMonth)

    while (newMonth < 0 ) newMonth = 12 + newMonth
      
    //month rollover
    if ( dates.month(date) !== ( newMonth % 12))
      date = dates.date(date, 0) //move to last of month

    return date
}

function createAccessor(method){
  return function(date, val){
    if (val === undefined)
      return date['get' + method]()

    date = new Date(date)
    date['set' + method](val)
    return date
  }
}

function createComparer(operator) {
  return function (a, b, unit) {
    return operator(+dates.startOf(a, unit), +dates.startOf(b, unit))
  };
}



/*****************
 ** WEBPACK FOOTER
 ** ./~/date-arithmetic/index.js
 ** module id = 415
 ** module chunks = 0
 **/
```

**Comment by jquense** [comment]
More then likely something isn't a Date object that should be. It may be your event start and end dates

**Comment by michaelimstepf** [comment]
You were correct, thanks a lot, you saved me a lot of time.

**Comment by HriBB** [comment]
I had the exact same problem. Although the docs for [startAccessor](http://intljusticemission.github.io/react-big-calendar/examples/index.html#prop-startAccessor) clearly say `The start date/time of the event. Must resolve to a JavaScript Date object.`, I totally missed it. Had to set a breakpoint and debug. Maybe we could do a simple check and throw an error if start/end is not a `Date`. Would probably prevent many issues like this.

**Comment by zelias500** [comment]
Along this line of thinking, what are people's thoughts on allowing the calendar to accept moment.js objects? Moment.js is already recommended as a localization option, so it's likely that most consumers of the component are making some use of moment already.

**Comment by jquense** [comment]
@zelias500 We are intentionally not coupling the calendar to any particular date library. Doing so would create a hard dependency on Moment, and internally it's just going to be converting them to normal Date objects anyway

**Comment by silentlight** [comment]
@michaelimstepf I am wondering how did you get it working. I am getting the same error. In my case I pull events from Rails API and my start and end dates are in format `Tue, 10 Jan 2017 17:33:00 UTC +00:00`. Do I have to iterate over all events and convert start and end dates to `Date`?

**Comment by michaelimstepf** [comment]
@silentlight Sorry I can't remember and haven't touched that codebase for a long time =(

**Comment by silentlight** [comment]
no worries @michaelimstepf :) @jdeniau do you know what I might be missing?

**Comment by jdeniau** [comment]
Hi, I do not remember the code well too, but it seems like you have a string, and the library requires a `Date` object.

You can just do something like this : `new Date('Tue, 10 Jan 2017 17:33:00 UTC +00:00')` and it should be fine

**Comment by DonovanCharpin** [comment]
Hey guys,

Are you sure the calendar is still compatible with moment? I have exactly the same error at the same line working with moment dates.

I initialized my calendar with
```
BigCalendar.momentLocalizer(moment); // or globalizeLocalizer
```
and the my calendar
```
<BigCalendar
    selectable={true}
    events={[{
         'title': 'My event',
         'start': moment(),
         'end': moment().add(4, 'hours')
    }]}
    startAccessor='start'
    endAccessor='end'
/>
```
The issue is that the method to get hours is get + method which is in my case `method="Hours"` so getHours()
```
function createAccessor(method){
  return function(date, val){
    if (val === undefined)
      return date['get' + method]()

    date = new Date(date)
    date['set' + method](val)
    return date
  }
}
```
But with moment, this method doesn't exist, the method is `moment().hours()` to get the number of hours.

**Comment by DonovanCharpin** [comment]
Oh my bad, the calendar expects only Date objects so moment().toDate()
```
events={[{
  'title': 'My event',
  'allDay': true,
  'start': moment().toDate(),
  'end': moment().add(4, "hours").toDate()
}]}
```

**Comment by tomahawkzx** [comment]
thanks @DonovanCharpin, this saved me some minutes

**Comment by steodor** [comment]
So yeah, for anyone having such issues in the future, here's the gist of it:

1. when using `moment.js` as your event dates, you need to go through them and `.toDate()` them all before passing them to the calendar
2. if you have strings like @silentlight then go through them all and either `new Date(<your date string>)` like @jdeniau said or `moment(<your date string>, [format]).toDate()` if it's in a format `new Date()` cannot understand

Basically the calendar only accepts javascript `Date` objects, not strings or `moment` objects, so make sure that's what you give it. This is intended behavior and it's fine because it makes the calendar library-agnostic, thus avoiding unnecessary dependencies. 

I agree with @HriBB on that a type check and a specific error message would improve overall usability of this library.

**Comment by fawadshahdev** [comment]
> Oh my bad, the calendar expects only Date objects so moment().toDate()
> 
> ```
> events={[{
>   'title': 'My event',
>   'allDay': true,
>   'start': moment().toDate(),
>   'end': moment().add(4, "hours").toDate()
> }]}
> ```

Hi , 
I am getting same error , could you please guide me where I can change `toDate` in code

**Comment by fawadshahdev** [comment]
> So yeah, for anyone having such issues in the future, here's the gist of it:
> 
>     1. when using `moment.js` as your event dates, you need to go through them and `.toDate()` them all before passing them to the calendar
> 
>     2. if you have strings like @silentlight then go through them all and either `new Date(<your date string>)` like @jdeniau said or `moment(<your date string>, [format]).toDate()` if it's in a format `new Date()` cannot understand
> 
> 
> Basically the calendar only accepts javascript `Date` objects, not strings or `moment` objects, so make sure that's what you give it. This is intended behavior and it's fine because it makes the calendar library-agnostic, thus avoiding unnecessary dependencies.
> 
> I agree with @HriBB on that a type check and a specific error message would improve overall usability of this library.

@steodor  please help me

**Comment by steodor** [comment]
@fawadshahdev sure, no problem. Give us the code here and we'll help.

**Comment by PeterFarnsworth23** [comment]
Has anyone seen this happen when trying to use the drag and drop? the events I am draginng and dropping are pulled from a postgres sql database and look like this.

[
    {
        "Id": 1,
        "title": "manualtest",
        "start": "Wed Aug 24 2019 1:00:00 GMT-0600",
        "end": "Wed Aug 24 2019 10:00:00 GMT-0600",
        "userId": 17,
        "description": "this is just a test"
    },
    {
        "Id": 4,
        "title": "crunch Time",
        "start": "Wed Aug 29 2019 18:00:00 GMT-0600",
        "end": "Wed Aug 29 2019 21:00:00 GMT-0600",
        "userId": 17,
        "description": "this is for fun"
    },
    {
        "Id": 13,
        "title": "what happened",
        "start": "Wed Aug 31 2019 18:00:00 GMT-0600",
        "end": "Wed Aug 31 2019 21:00:00 GMT-0600",
        "userId": 17,
        "description": null
    },]

the Start and end dates were created using moment and show on the calendar on render. The problem is as soon as I try to drag the event I get this "Uncaught TypeError: d[("get" + method)] is not a function index.js.242".

I am not sure why it can read on the render but not the default drag and drop. 

Thank you

**Comment by buithienquyet** [comment]
Has same problem when using addon dragAndDrop

**Comment by steodor** [comment]
Sorry guys, i'm not working on anything involving this project any more, and my time is full so i won't be able to reproduce/help. But simply put, try debugging, go through the stack trace carefully and try to pinpoint at which point the start/end dates are getting broken (i.e. no longer `Date` instances). The error itself is simple, and debugging should be too - you just need a few minutes of silence and focus to go through it methodically. If you ensure all the dates are being converted into `Date` and remain that way, you should be fine. Cheers!

**Comment by buithienquyet** [comment]
Thank you very much! I have fixed this problem, I am so sorry because I did
not check it carefully before asking question, again, thanks!

Vào 15:16, Th 3, 1 thg 10, 2019 Teodor Sandu <notifications@github.com> đã
viết:

**Comment by ozzy432836** [comment]
Thanks guys. This really helped me. I was doing moment().add(1,"hour").format("YYYY-MM-DD HH:mm") which I found is not a JS date object.

If this helps anyone here is a quick BigCalendar with some test dates

```
const BigCalendarComponent = ({ stateProps, dispatchProps }) => {
  const localizer = momentLocalizer(moment);

  let now = stateProps.currentDate === "" ? moment().format('YYYY-MM-DD') : stateProps.currentDate;
  
  let d1Start = moment().add(1, "hours").toDate();
  let d1End = moment(d1Start).add(1, "hours").toDate();

  let d2Start = moment().add(3, "hours").toDate();
  let d2End = moment(d2Start).add(1, "hours").toDate();

  let d3Start = moment().add(6, "hours").toDate();
  let d3End = moment(d2Start).add(1, "hours").toDate();

  console.log(d1Start);

  var isDate = Object.prototype.toString.call(d1Start) === '[object Date]';
  console.log(isDate);

  var d = new Date();
  console.log(d);
  isDate = Object.prototype.toString.call(d) === '[object Date]';
  console.log(isDate);

  var events = [
    { id: 0, title: "My Event", start: d1Start, end: d1End },
    { id: 1, title: "My Event2", start: d2Start, end: d2End },
    { id: 2, title: "My Event2", start: d3Start, end: d3End },
  ];

  return (
    <div style={{ height: "100%" }}>
      <Header>Schedule Manager</Header>
      <Calendar
        localizer={localizer}
        events={events}
        startAccessor="start"
        endAccessor="end"
      />
    </div>
  );
}
```

**Comment by jacklimwenjie** [comment]
I'm wondering if this parsing the time could be part of the default handling of the lib. 

I mean a localizer is already needed to be specified when using this lib, so it's technically feasible to parse the event object `start` and `end` value based on the localizer. 

For example, if `moment` is chosen as the localizer, passing `start` value as `"2019-11-11T00:00:00Z"` should automatically parsed by using `moment("2019-11-11T00:00:00Z").toDate()`.

**Comment by androBermudez** [comment]
I am facing this issue when setting min time and rendering week view. It works in Day view but throw an exception when navigate to week.

**Comment by hechleR48** [comment]
Yes i have the same problem in week view. My month view is rendering and a can navigate but. If a changed to week an navigate to a date i got the error: "Uncaught TypeError: d[("get" + method)] is not a function" 
Any ideas?

**Comment by shallonkobusinge** [comment]
Make sure that you have the correct values for start and end keys in your event object, your event object should be like :

```
data = [
  {
    title: "My event",
    allDay: false,
    start: new Date(2020, 10, 25, 10, 0), // 10.00 AM
    end: new Date(2020, 10, 25, 11, 0), // 2.00 PM
  }
```

**Comment by adisaktijrs** [comment]
After I read this issue, I finally got the answer. I'll leave my solution here, in case there are some guys who faced the same issue when working with this lib and need to store the data to local storage in a React app.

I faced a problem when storing the 'raw' `start` and `end` Date data to local storage. So I fixed it by reformating the date data to a **timestamp** and **store it in local storage**. Then when I need to read it back, I'm reformating it back **to date data.**

In a componentDidUpdate() function, when I stored the data to local storage:
```
const newEvents = this.state.events.map((event) => {
  return {
    id: event.id,
    start: new Date(event.start).getTime(),
    end: new Date(event.end).getTime(),
    title: event.title,
  };
});
const events = JSON.stringify(newEvents);
localStorage.setItem("events", events);
```

and in a componentDidMount I reformating it to a js Date:
```
const events = localStorage.getItem("events");
const result = JSON.parse(events);
const newEvents = result.map((event) => {
  return {
    id: event.id,
    start: new Date(event.start),
    end: new Date(event.end),
    title: event.title,
  };
});
```

I'm not too sure whether is it a good solution or not, at least it works just fine for now. It's midnight here, and I need to sleep🤞

**Comment by MhamedMonastiri** [comment]
> moment

i have the same problem ,
it was working fine until today , the week view doesn't work anymore.
i didn't change nothing and the month , day and calendar view works fine...
